### PR TITLE
Leaves - Cloudy and Julia K

### PR DESCRIPTION
**Pull request by Kalakalot**
# Ride Share
Congratulations! You're submitting your assignment!

## Comprehension Questions
Question | Answer
:------------- | :-------------
How did getting up to speed on the existing codebase go? If it went well, what worked? If it didn't go well, what will you do differently next time? | Oh geez, it took us a LONG time to get up to speed on the codebase, even though we mapped out the relationships and answered the comprehension questions before diving in. There were multiple updates to the starter codebase that we didn't catch until spending an hour or more trying to figure out why we kept getting errors, which was frustrating and made us doubt our ability as programmers. Next time we will be checking project notification FREQUENTLY for updates to the codebase and will also spend even longer getting familiar with how everything fits together. 
What _inheritance_ relations exist between classes? | Trip, Passenger and Driver all inherit from their parent CSVrecord. 
What _composition_ relations exist between classes? | Trip Dispatcher is composed of Trips, Passengers, and Drivers. Trip is composed of Passengers and Trips. Passenger is composed of Trips. Driver is composed of Trips. 
Describe a decision you had to make when working on this project. What options were you considering? What helped you make your final decision? | Initially we added an error message for situations where the passenger net expenditures method was being called on a passenger who hadn't taken any trips. But when we thought about it, we realized that "no trips yet" isn't really an error or an unusual situation, after all, every passenger is going to start out as not having taken any trips. We figured that receiving an error message in this situation would be a poor user experience, so instead passenger net expenditures just returns 0 if the passenger hasn't taken any trips. 
Give an example of a _template method_ that you implemented for this assignment | We created the driver class from the CsvRecord template. 
Give an example of a _nominal test_ that you wrote for this assignment | The test for total revenue (driver_test: "returns net revenue for a driver") provides standard driver and trip data and the expected result (no funny business here). 
Give an example of an _edge case test_ that you wrote for this assignment | Since the specifications for the way driver revenue is calculated (company takes a $1.65 fee per trip, no minimum trip cost) could result in the driver earning negative revenue, we added a test that checks that the expected total is returned for negative earnings. (Although the fact that a driver could generate negative earnings seems really unfair.)
What is a concept that you gained more clarity on as you worked on this assignment | We knew that the steps for TDD are supposed to be a) write a test you know will fail, b) make your code pass the test, then c) refactor your code, but, in practice, most of our tests before and for the first part of this project were a) write code, b) write a test, c) stop there. During wave 3, we got better at writing tests first and then writing code to pass the tests. Next up: refactoring code. (It was a real stretch just getting tests to pass and we ran out of time to do any refactoring.)

### CODE REVIEW

File: test/passenger_test.rb [hunk 1]
```diff
@@ -1,40 +1,40 @@
...
       expect(@passenger.trips).must_be_kind_of Array
     end
   end
-
-
+  
+  
   describe "trips property" do
     before do
       # TODO: you'll need to add a driver at some point here.
```
Comment (dHelmgren) [inline test/passenger_test.rb:40]: Looks like you missed this. It's always good to do a `command + shift + f` for TODO before you commit.

File: test/trip_dispatch_test.rb [hunk 2]
```diff
@@ -8,118 +9,166 @@ def build_test_dispatcher
...
-        end
-      end
+    it "passenger is an instance of Passenger" do
+      expect(@trip.passenger).must_be_kind_of RideShare::Passenger
     end
-  end
+    
+    it "adds the new trip to the passenger's list of trips" do
+    expect @trip.passenger.trips.include?(@trip)
+    end
```
Comment (dHelmgren) [inline test/trip_dispatch_test.rb:171]: Very thorough tests! Nice work!

**Comment by dHelmgren** [comment]
# Ride Share
## What We're Looking For

Feature|Feedback
:------------- | :-------------
**Baseline** |
Used Git Regularly | yes
Answer comprehension questions | for: `Give an example of a template method that you implemented for this assignment`: Your driver class isn't a template method (it's a class), but its `from_csv` is.
**Wave 1** |
Appropriate use of Ruby's `Time` | yes
`Trip` has a helper method to calculate duration | yes
`Passenger` methods to calculate total cost and total time | yes
Complete tests for new behavior |
**Wave 2** |
`Driver` inherits from `CsvRecord`, and implements `from_csv` | yes
`Driver`s and `Trip`s are connected by `TripDispatcher` | yes
`Driver` has method to calculate average rating | yes
`Driver` has a method to calculate total revenue | yes
**Wave 3** |
`TripDispatcher` has a new method to create trips | yes
Trip creation logic is correctly implemented | yes
New `Trip` is connected to the appropriate `Driver` and `Passenger` | yes
Tests for `request_trip` | yes
Wave 1 and 2 methods ignore in-progress trips | yes
Tests for Wave 1 and 2 methods w/ in-progress trips | I don't see you testing them with the new in-progress trips. Any of the methods from waves 1 & 2 that do nil checking for trips should be tested with at least one in-progress trip.
Appropriate helper methods were made to help with complex logic | yes
**Overall** | Solid work on this! Don't forget to go back and update tests when you add new functionality to old classes!
